pkg/controllers/openshift-apiserver.go

```diff
@@ -32,48 +33,91 @@ import (
-	cmd := &cobra.Command{
-		Use:   "openshift-apiserver",
-		Short: "Command for the OpenShift API Server",
-		Run: func(cmd *cobra.Command, args []string) {
-			cmd.Help()
-			os.Exit(1)
-		},
-	}
-	start := openshift_apiserver.NewOpenShiftAPIServerCommand("start", os.Stdout, os.Stderr, stopCh)
-	cmd.AddCommand(start)
+type OCPAPIServer struct {
+	ConfigFile string
+	Output     io.Writer
+	cfg        *config.MicroshiftConfig
+}
 
-	return cmd
+func NewOpenShiftAPIServer(cfg *config.MicroshiftConfig) *OCPAPIServer {
+	s := &OCPAPIServer{}
+	s.configure(cfg)
+	return s
 }
 
-func OCPAPIServer(cfg *config.MicroshiftConfig) error {
-	stopCh := make(chan struct{})
-	command := newOpenshiftApiServerCommand(stopCh)
+func (s *OCPAPIServer) Name() string           { return "ocp-apiserver" }
+func (s *OCPAPIServer) Dependencies() []string { return []string{"kube-apiserver"} }
+
+func (s *OCPAPIServer) configure(cfg *config.MicroshiftConfig) error {
+	var configFilePath = cfg.DataDir + "/resources/openshift-apiserver/config/config.yaml"
+
+	if err := OpenShiftAPIServerConfig(cfg); err != nil {
+		logrus.Infof("Failed to create a new ocp-apiserver configuration: %v", err)
+		return err
+	}
 	args := []string{
-		"start",
-		"--config=" + cfg.DataDir + "/resources/openshift-apiserver/config/config.yaml",
-		"--authorization-kubeconfig=" + cfg.DataDir + "/resources/kubeadmin/kubeconfig",
-		"--authentication-kubeconfig=" + cfg.DataDir + "/resources/kubeadmin/kubeconfig",
-		"--logtostderr=" + strconv.FormatBool(cfg.LogDir == "" || cfg.LogAlsotostderr),
+		"--config=" + configFilePath,
 		"--alsologtostderr=" + strconv.FormatBool(cfg.LogAlsotostderr),
 		"--v=" + strconv.Itoa(cfg.LogVLevel),
 		"--vmodule=" + cfg.LogVModule,
+		"--logtostderr=" + strconv.FormatBool(cfg.LogDir == "" || cfg.LogAlsotostderr),
+	}
+
+	options := openshift_apiserver.OpenShiftAPIServer{Output: os.Stdout}
+	options.ConfigFile = configFilePath
+
+	cmd := &cobra.Command{
+		Use:          "ocp-apiserver",
+		Long:         "ocp-apiserver",
+		SilenceUsage: true,
+		RunE:         func(cmd *cobra.Command, args []string) error { return nil },
 	}
-	if cfg.LogDir != "" {
-		args = append(args, "--log-dir="+cfg.LogDir)
+
+	cmd.SetArgs(args)
+	cmd.MarkFlagFilename("config", "yaml", "yml")
+	cmd.MarkFlagRequired("config")
+	cmd.ParseFlags(args)
+
+	s.ConfigFile = options.ConfigFile
+	s.Output = options.Output
+	s.cfg = cfg
+
+	err := s.prepareOCPComponents(s.cfg)
+	if err != nil {
+		logrus.Warningf("Failed to start ocp-apiserver %v", err)
 	}
+	logrus.Infof("%s is ready", s.Name())
+
+	return nil
+
+}
+
+func (s *OCPAPIServer) Run(ctx context.Context, ready chan<- struct{}, stopped chan<- struct{}) error {
+	defer close(stopped)
 
-	command.SetArgs(args)
-	logrus.Infof("starting openshift-apiserver, args: %v", args)
-	go func() {
-		logrus.Fatalf("ocp apiserver exited: %v", command.Execute())
-	}()
+	// go func() {
+	// 	err := s.prepareOCPComponents(s.cfg)
```

Comment:
you can refactor `prepareOCPComponents` and move `waitForOCPAPIServer` to here as health checker